### ISSUE TITLE
LogPublisher's observer error handling can publish events to a different LogPublisher

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3307]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3307|
|Type|defect|
|Created|2008-06-23 20:33:51Z|
|Branch|https://github.com/twisted/twisted/tree/non-global-publisher-err-3307|

`LogPublisher` uses the global `err` when an observer raises an exception in `msg`.  Only for the global `LogPublisher` will this result in the event being sent to back to itself.  This makes testing annoying and is slightly surprising.  `LogPublisher` should send error reports to itself, not something else.

Attachments:

* [report-observe-errors-privately-3307.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/24c/24cd63144a915209ec5d4f0270802326e3212121/b4a9437000c43122ebc2815b2b69d0174e85644a.patch) (3450 bytes) - added by zseil on 2011-07-12 23:13:21Z - 
* [report-observe-errors-privately-3307-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/24c/24cd63144a915209ec5d4f0270802326e3212121/5cdd112309e987d4e602971b2106e6537f6438a3.patch) (6174 bytes) - added by zseil on 2011-07-13 22:11:50Z - updated patch

<details><summary>Searchable metadata</summary>

```
trac-id__3307 3307
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_non_global_publisher_err_3307 branches/non-global-publisher-err-3307
branch_author__zseil zseil
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1214253231000000 1214253231000000
changetime__1313283701000000 1313283701000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_2"></a>|@zseil commented|
|-|-|

Attached is a minimal patch that tries to fix this issue by adding a private `_err` method to `LogObservers`, a stripped down version of the global `log.err`, and uses it to report broken observers. The patch was separated from patch for [#5194](https://github.com/twisted/twisted/issues/5194), the changes that were not strictly needed were removed.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_3"></a>|@zseil commented|
|-|-|

Ups, I missed that my change makes one of the tests useless, I noticed it only after a Pyflakes run. Updated patch attached, I converted the `test_doubleErrorDoesNotRemoveObserver` to check the new `_err` method of observers, fixed the Pyflakes complaints and added a check that should catch errors like mine in the future.

Sorry for the churn, I'll try to keep my requests for review less of a moving target in the future.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32451]) Branching to 'non-global-publisher-err-3307'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32452]) Apply report-observe-errors-privately-3307-2.patch 

refs [#3307](https://github.com/twisted/twisted/issues/3307)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks for working on this zseil.  Sorry about the delay in the review.  Don't worry about the churn, it doesn't appear to have cost anybody any time. :)

This patch looks excellent.  I don't have any suggestions for changes.  [Buildbot is also happy with it.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/non-global-publisher-err-3307)  I will merge to trunk immediately.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [32453]) Merge non-global-publisher-err-3307

Author: zseil
Reviewer: exarkun
Fixes: [#3307](https://github.com/twisted/twisted/issues/3307)

When an observer on `LogPublisher` raises an exception, broadcast that
exception event to the `LogPublisher`'s own observers, instead of to
the global `LogPublisher` instances observers.
